core/src/main/java/org/apache/calcite/rel/rules/MinusToFilterRule.java

```diff
@@ -1,121 +0,0 @@
-/*
- * Licensed to the Apache Software Foundation (ASF) under one or more
- * contributor license agreements.  See the NOTICE file distributed with
- * this work for additional information regarding copyright ownership.
- * The ASF licenses this file to you under the Apache License, Version 2.0
- * (the "License"); you may not use this file except in compliance with
- * the License.  You may obtain a copy of the License at
- *
- * http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-package org.apache.calcite.rel.rules;
-
-import org.apache.calcite.plan.RelOptRuleCall;
-import org.apache.calcite.plan.RelRule;
-import org.apache.calcite.rel.RelNode;
-import org.apache.calcite.rel.core.Filter;
-import org.apache.calcite.rel.core.Minus;
-import org.apache.calcite.rex.RexNode;
-import org.apache.calcite.rex.RexUtil;
-import org.apache.calcite.tools.RelBuilder;
-
-import org.immutables.value.Value;
-
-/**
- * Rule that replaces {@link Minus} operator with {@link Filter}
- * when both inputs are from the same source with only filter conditions differing.
- * Only inspect a single {@link Filter} layer in the inputs of {@link Minus}.
- * For inputs with nested {@link Filter}s, apply {@link CoreRules#FILTER_MERGE}
- * as a preprocessing step.
- *
- * <p>Example transformation:
- * <blockquote><pre>
- * SELECT mgr, comm FROM emp WHERE mgr = 12
- * EXCEPT
- * SELECT mgr, comm FROM emp WHERE comm = 5
- *
- * to
- *
- * SELECT DISTINCT mgr, comm FROM emp
- * WHERE mgr = 12 AND NOT(comm = 5)
- * </pre></blockquote>
- */
-@Value.Enclosing
-public class MinusToFilterRule
```

Comment:
I don't think you can delete a public class.
Just mark it `@deprecated` and point to the replacement.
I know, this is annoying.

Reply:
Oh, I suddenly understand what you mean. I don't want to delete the MinusToFilterRule class, but mark it as obsolete. Then I don't want to delete MINUS_TO_FILTER in CoreRules. The new SetOpToFilterRule is completely new. So the name here should be written as MINUS_FILTERTO_FILTER, and the others are similar. Do I understand it correctly now?

Reply:
That sounds like a good approach, but I would have a _ before the TO.

Reply:
Sorry, it was my writing mistake. It should be MINUS_FILTER_TO_FILTER.